heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerRunner.java

```diff
@@ -196,4 +246,71 @@ protected boolean cleanState(
     LOG.fine("Cleaned up topology state");
     return true;
   }
+
+  @VisibleForTesting
+  PackingPlans.PackingPlan buildNewPackingPlan(PackingPlans.PackingPlan currentProtoPlan,
+                                               Map<String, Integer> changeRequests,
+                                               TopologyAPI.Topology topology) {
+    PackingPlanProtoDeserializer deserializer = new PackingPlanProtoDeserializer();
+    PackingPlanProtoSerializer serializer = new PackingPlanProtoSerializer();
+    PackingPlan currentPackingPlan = deserializer.fromProto(currentProtoPlan);
+
+    Map<String, Integer> componentCounts = currentPackingPlan.getComponentCounts();
+    Map<String, Integer> componentChanges = parallelismDelta(componentCounts, changeRequests);
+
+    // Create an instance of the packing class
+    String repackingClass = Context.repackingClass(config);
+    IRepacking packing;
+    try {
+      // create an instance of the packing class
+      packing = ReflectionUtils.newInstance(repackingClass);
+    } catch (IllegalAccessException | InstantiationException | ClassNotFoundException e) {
+      throw new IllegalArgumentException(
+          "Failed to instantiate packing instance: " + repackingClass, e);
+    }
+    try {
+      packing.initialize(config, topology);
+      PackingPlan packedPlan = packing.repack(currentPackingPlan, componentChanges);
+      return serializer.toProto(packedPlan);
+    } finally {
+      SysUtils.closeIgnoringExceptions(packing);
+    }
+  }
+
+  @VisibleForTesting
+  Map<String, Integer> parallelismDelta(Map<String, Integer> componentCounts,
+                                        Map<String, Integer> changeRequests) {
+    for (String component : changeRequests.keySet()) {
+      if (!componentCounts.containsKey(component)) {
+        throw new IllegalArgumentException(
+            "Invalid component name in update request: " + component);
+      }
+      Integer newValue = changeRequests.get(component);
+      Integer delta = newValue - componentCounts.get(component);
+      if (delta == 0) {
+        changeRequests.remove(component);
+      } else {
+        changeRequests.put(component, delta);
+      }
+    }
+    return changeRequests;
```

Comment:
The input request Map is modified by this method. I think it would be better to create a new structure with the delta to avoid any confusion.